## Setting enviroment

In [94]:
# !pip install -q datasets==3.6.0

In [95]:
import datasets

print(datasets.__version__)

3.6.0


## Que 1

In [96]:
from datasets import get_dataset_config_names

configs = get_dataset_config_names("ai4bharat/naamapadam")

print(configs)
print("Total configs:", len(configs))

['as', 'bn', 'gu', 'hi', 'kn', 'ml', 'mr', 'or', 'pa', 'ta', 'te']
Total configs: 11


In [97]:
from datasets import load_dataset

hi_ds = load_dataset("ai4bharat/naamapadam", "hi")

hi_ds

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 985787
    })
    test: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 867
    })
    validation: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 13460
    })
})

In [98]:
from datasets import load_dataset

ds = load_dataset("ai4bharat/naamapadam", "ta", split='train+test+validation')

ds

Dataset({
    features: ['tokens', 'ner_tags'],
    num_rows: 501435
})

In [99]:
ds.cache_files

[{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-train.arrow'},
 {'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-test.arrow'},
 {'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-validation.arrow'}]

In [100]:
ds.features

{'tokens': Sequence(feature=Value(dtype='string', id=None), length=-1, id=None),
 'ner_tags': Sequence(feature=ClassLabel(names=['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC'], id=None), length=-1, id=None)}

In [101]:
len(ds.features['ner_tags'].feature.names)

7

## Que 2

In [102]:
from datasets import load_dataset

ds = load_dataset("ai4bharat/naamapadam", "ta")

ds

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 497882
    })
    test: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 758
    })
    validation: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 2795
    })
})

In [103]:
ds.cache_files

{'train': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-train.arrow'}],
 'test': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-test.arrow'}],
 'validation': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-validation.arrow'}]}

## Que 3

`ds.cache_files`: represents the size of the downloaded dataset stored locally in the cache.

In [104]:
from datasets import load_dataset
import os

ds = load_dataset("ai4bharat/naamapadam", "ta")

total_size = 0

for split in ds.cache_files:
    for file_info in ds.cache_files[split]:
        size_mb = os.path.getsize(file_info["filename"]) / (1024 * 1024)
        total_size += size_mb

print(f"Total dataset size: {total_size:.2f} MB")

Total dataset size: 179.29 MB


### `load_dataset_builder()`

- Does not download or load the dataset itself
- Creates a dataset builder object containing metadata and download instructions
- Lets you inspect information about the dataset before downloading it

In [105]:
from datasets import load_dataset_builder

builder = load_dataset_builder("ai4bharat/naamapadam", "ta")


In [106]:
builder.info.features

{'tokens': Sequence(feature=Value(dtype='string', id=None), length=-1, id=None),
 'ner_tags': Sequence(feature=ClassLabel(names=['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC'], id=None), length=-1, id=None)}

In [107]:
builder.info.description

'\n'

The total size of the files that need to be downloaded from the source.

- Think of it as, If you download a ZIP file from the internet: `dataset.zip  → 50 MB`
- then download_size is the size of that ZIP (or source files).

In [108]:
print("Download size:",
      builder.info.download_size / (1024*1024), "MB")

Download size: 46.37075233459473 MB


The size of the dataset after it has been processed and stored in Hugging Face's Arrow format.

- Downloaded files (50 MB)

        ↓

- Processed by datasets library

        ↓

- Arrow files (100 MB)

In [109]:
print("Dataset size:",
      builder.info.dataset_size / (1024*1024), "MB")

Dataset size: 179.84503650665283 MB


In [110]:
builder.info.splits

{'train': SplitInfo(name='train', num_bytes=187209743, num_examples=497882, shard_lengths=None, dataset_name='naamapadam'),
 'test': SplitInfo(name='test', num_bytes=320938, num_examples=758, shard_lengths=None, dataset_name='naamapadam'),
 'validation': SplitInfo(name='validation', num_bytes=1050508, num_examples=2795, shard_lengths=None, dataset_name='naamapadam')}

## Quick Summary

| Attribute | What it Represents | When Measured |
|------------|-------------------|----------------|
| `builder.info.download_size` | Size of the files downloaded from the dataset source (compressed/raw files) | Before processing |
| `builder.info.dataset_size` | Size of the processed dataset stored in Apache Arrow format | After processing |
| `os.path.getsize()` on `ds.cache_files` | Actual size of the cached dataset files stored on your machine | After download and processing |

### In Simple Terms

```text
Source Files (download_size)
            ↓
      Processing
            ↓
Arrow Dataset Files (dataset_size)
            ↓
Stored in Cache Directory
            ↓
Actual File Size on Disk (os.path.getsize)

## Que 4

In [111]:
ds['train'].num_rows

497882

## Que 5

In [112]:
ds['train'][0]

{'tokens': ['பைரவருக்கு',
  'தேய்பிறை',
  'அஷ்டமியில்',
  'விசேஷ',
  'அபிஷேக',
  'ஆராதனைகள்',
  'நடைபெறுகின்றன',
  '.'],
 'ner_tags': [1, 0, 0, 0, 0, 0, 0, 0]}

Creating `num tokens` column

In [113]:
def add_token_col(example):
    example['num tokens'] = len(example['tokens'])
    
    return example

In [114]:
ds = ds.map(add_token_col)

ds

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags', 'num tokens'],
        num_rows: 497882
    })
    test: Dataset({
        features: ['tokens', 'ner_tags', 'num tokens'],
        num_rows: 758
    })
    validation: Dataset({
        features: ['tokens', 'ner_tags', 'num tokens'],
        num_rows: 2795
    })
})

In [115]:
ds['train'][0]['num tokens'] == len(ds['train'][0]['tokens'])


True

calculating the total number of columns in the entire dataset

In [116]:
total_tokens = 0

for split in ds:
    total_tokens += sum(ds[split]['num tokens'])
    
total_tokens/1_000_000

6.001876

## Que 6

In [117]:
ds.cache_files

{'train': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-339a5c9b5e072e5d.arrow'}],
 'test': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-4163f04c34e92fd8.arrow'}],
 'validation': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-c572dfa9f0baec96.arrow'}]}

In [118]:
size = 0

for split in ds.cache_files:
    for file_info in ds.cache_files[split]:
        size += os.path.getsize(file_info['filename']) /(1024*1024)

print(f'Total_modified_size: {size} MB')

Total_modified_size: 183.1378173828125 MB


- The modified dataset itself occupies approximately 183 MB, only slightly larger than the original dataset (~179 MB). However, because Hugging Face datasets are immutable and retain the original cache files, the total cache storage consumed on disk becomes roughly 179 MB + 183 MB ≈ 362 MB, which is slightly more than double the original dataset size.

## Que 7

In [119]:
from datasets import concatenate_datasets

new_ds = concatenate_datasets(
    [
            ds['train'],
    ds['test'],
    ds['validation']
    ],
    axis=0
)

new_ds

Dataset({
    features: ['tokens', 'ner_tags', 'num tokens'],
    num_rows: 501435
})

In [120]:
def add_text_col(example):
    example['text'] = ' '.join(example['tokens'])
    
    return example

In [121]:
new_ds = new_ds.map(add_text_col)

new_ds[0]

{'tokens': ['பைரவருக்கு',
  'தேய்பிறை',
  'அஷ்டமியில்',
  'விசேஷ',
  'அபிஷேக',
  'ஆராதனைகள்',
  'நடைபெறுகின்றன',
  '.'],
 'ner_tags': [1, 0, 0, 0, 0, 0, 0, 0],
 'num tokens': 8,
 'text': 'பைரவருக்கு தேய்பிறை அஷ்டமியில் விசேஷ அபிஷேக ஆராதனைகள் நடைபெறுகின்றன .'}

In [122]:
ds = new_ds.remove_columns(['ner_tags', 'tokens'])

In [123]:
ds.column_names

['num tokens', 'text']

In [124]:
ds.num_rows

501435

In [125]:
ds

Dataset({
    features: ['num tokens', 'text'],
    num_rows: 501435
})

## Que 8


### Why Dataset Modifications Increase Disk Usage

Hugging Face datasets are **immutable**. Operations such as:

```python
ds = ds.map(...)
ds = ds.filter(...)
ds = ds.remove_columns(...)
ds = ds.rename_column(...)
```

do **not** modify the existing Apache Arrow file in place.

Instead, the library:

1. Reads the existing cached dataset.
2. Creates a **new Arrow cache file** containing the modified dataset.
3. Keeps the old cache file for reproducibility and caching purposes.

As a result, the cache directory may contain multiple versions of the dataset, causing disk usage to increase over time.

#### Example

Initial dataset:

```text
cache/
├── original.arrow
```

After:

```python
ds = ds.map(add_token_col)
```

Cache directory:

```text
cache/
├── original.arrow
├── modified.arrow
```

After another transformation:

```python
ds = ds.filter(...)
```

Cache directory:

```text
cache/
├── original.arrow
├── modified.arrow
├── filtered.arrow
```

#### Conclusion

Each dataset transformation typically creates a new cache file rather than modifying the existing one. Therefore, repeated operations such as `map()`, `filter()`, and `remove_columns()` can gradually increase the disk space consumed by the Hugging Face cache.

## Que 9

In [126]:
ds = ds.filter(lambda row: row['num tokens'] >= 6)

In [127]:
ds

Dataset({
    features: ['num tokens', 'text'],
    num_rows: 370495
})

## Que 10

In [128]:
tamil_ds = load_dataset('ai4bharat/indic_glue', 'inltkh.ta', split='train+test+validation')

tamil_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 6684
})

In [129]:
tamil_ds = tamil_ds.rename_column("text", "num_tokens")

tamil_ds 

Dataset({
    features: ['num_tokens', 'label'],
    num_rows: 6684
})

In [131]:
tamil_ds = tamil_ds.filter(
    lambda row: len(row["num_tokens"].split(' ')) >= 6
)

Filter: 100%|██████████| 6684/6684 [00:00<00:00, 150069.20 examples/s]


In [132]:
tamil_ds

Dataset({
    features: ['num_tokens', 'label'],
    num_rows: 6428
})

In [ ]:
# tamil_ds = tamil_ds.remove_columns('label')

In [ ]:
# tamil_ds['text'][0]

'கே.வி.ஆனந்தே ட்விட்டரில் இதை அறிவித்துள்ளார். இந்தப் படத்துக்கு கேவ்மிக் ஆரி ஒளிப்பதிவு செய்ய, ஹாரிஸ் ஜெயராஜ் இசையமைக்கிறார். பட்டுக்கோட்டை பிரபாகர் வசனம் எழுத, கலை இயக்குநராக கிரண் பணியாற்றுகிறார். இந்தப் படத்தை லைகா புரொடக்\u200cஷன்ஸ் நிறுவனம் தயாரிக்கிறது.'

In [ ]:
# def add_col(col):
#     col['num_tokens'] = len(col['text'].split(' '))
    
#     return col

In [ ]:
# tamil_ds = tamil_ds.map(add_col)

In [ ]:
tamil_ds

Dataset({
    features: ['text', 'num_tokens'],
    num_rows: 6213
})

In [133]:
from datasets import interleave_datasets

id_data = interleave_datasets(
    [
        ds,
        tamil_ds,
    ],
    probabilities=[0.8, 0.2],
    seed = 42
)

id_data

Dataset({
    features: ['num tokens', 'text', 'num_tokens', 'label'],
    num_rows: 32354
})

# Revisioning

In [ ]:
!ls -l ~/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc

total 1175152
-rw-rw-r-- 1 mypc mypc   1070552 Jun 13 13:22 cache-31fb67ca58fbaf0a.arrow
-rw-rw-r-- 1 mypc mypc 190636448 Jun 13 15:01 cache-339a5c9b5e072e5d.arrow
-rw-rw-r-- 1 mypc mypc    326920 Jun 13 15:01 cache-4163f04c34e92fd8.arrow
-rw-rw-r-- 1 mypc mypc 311373432 Jun 13 14:11 cache-446dfcae5978cc8e.arrow
-rw-rw-r-- 1 mypc mypc 190636448 Jun 13 13:22 cache-6b9b8ce642206145.arrow
-rw-rw-r-- 1 mypc mypc    326920 Jun 13 13:22 cache-8d21fe9878e7886b.arrow
-rw-rw-r-- 1 mypc mypc   3036552 Jun 14 17:28 cache-9aed762bcebd41a8.arrow
-rw-rw-r-- 1 mypc mypc   1070552 Jun 13 15:01 cache-c572dfa9f0baec96.arrow
-rw-rw-r-- 1 mypc mypc   2730224 Jun 13 14:27 cache-c67b46c9543d9fad.arrow
-rw-rw-r-- 1 mypc mypc   2730224 Jun 13 15:03 cache-ec95276cab5993af.arrow
-rw-rw-r-- 1 mypc mypc 311373432 Jun 13 15:01 cache-feb478bddb24dd1b.arrow
-rw-rw-r-- 1 mypc mypc      1377 Jun 13 11:50 dataset_info.json
-rw-rw-r-- 1 mypc mypc        75 Jun 13 11:50 LICENSE
-rw-rw-r-- 1 mypc mypc    320672 Jun 13 11: